### PR TITLE
✔xref resolving

### PR DESCRIPTION
**Pull request by live1206**
- No need to remove the host name while resolving xref
- Append `?branch=master` to xref URL while build branch is `master`

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4772)

### CODE REVIEW

File: src/docfx/build/xref/XrefMapBuilder.cs [hunk 1]
```diff
@@ -166,7 +166,7 @@ private static (List<Error> errors, IReadOnlyList<InternalXrefSpec> specs) LoadS
                 var xref = new InternalXrefSpec
                 {
                     Uid = item.Key,
-                    Href = isRoot ? file.CanonicalUrlWithoutLocale : $"{file.CanonicalUrlWithoutLocale}#{GetBookmarkFromUid(item.Key)}",
+                    Href = isRoot ? file.SiteUrl : $"{file.SiteUrl}#{GetBookmarkFromUid(item.Key)}",
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMapBuilder.cs]: remove `CanonicalUrlWithoutLocale` from document

File: docs/specs/moniker.yml [hunk 2]
```diff
@@ -445,8 +445,6 @@ outputs:
       "conceptual": "<p>Moniker: netcore-2.0</p>",
       "monikers": ["netcore-2.0"]
     }
-  xrefmap.json: | 
-    {"references":[{"uid":"a","href":"https://docs.com/docs/a","name":"netcore-2.0"}]}
```
Comment (OsmondJiang) [inline docs/specs/moniker.yml]: @928PJY is this for testing monikers?

File: docs/specs/xref.yml [hunk 3]
```diff
@@ -200,29 +202,31 @@ outputs:
...
+          }
 outputs:
   docs/a.json:
   docs/b.json: | 
     {
       "content": {
         "$schema": "https://raw.githubusercontent.com/dotnet/docfx/v3/schemas/TestData.json",
         "xref": {
-            "href": "https://docs.com/docs/a.json",
+            "href": "/docs/a.json",
```
Comment (OsmondJiang) [inline docs/specs/xref.yml]: why it's removed host?

File: src/docfx/build/xref/XrefMap.cs [hunk 4]
```diff
@@ -55,16 +55,15 @@ public void OutputXrefMap(Context context)
         }
 
         private (Error error, string href, string display, IXrefSpec xrefSpec) ResolveCore(
-            string uid, SourceInfo<string> href, string displayPropertyName, Document relativeTo, string moniker = null)
```
Comment (OsmondJiang) [inline src/docfx/build/xref/XrefMap.cs]: rebase add a test

File: docs/specs/xref.yml (hunk 3 above)
Comment (live1206) [reply docs/specs/xref.yml]: internal defined UID should be resolved to relative path

File: src/docfx/build/xref/XrefMapBuilder.cs (hunk 1 above)
Comment (live1206) [reply src/docfx/build/xref/XrefMapBuilder.cs]: it is still in use when outputting xref map

File: docs/specs/moniker.yml (hunk 2 above)
Comment (live1206) [reply docs/specs/moniker.yml]: we do not output moniker information in xref map for now

File: src/docfx/build/xref/XrefMap.cs (hunk 4 above)
Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: https://github.com/dotnet/docfx/blob/f8a8fca15ea624d1f53f8c773aacaf403220dfc1/docs/specs/xref.yml#L488
We have a test of `Markdown inclusion with xref reference`, the resolved URL is a relative path

**Comment by live1206** [comment]
The diff in the integration test is expected, appending `?branch=master` to xref map output

File: src/docfx/build/xref/InternalXrefSpec.cs [hunk 5]
```diff
@@ -29,19 +29,36 @@ public string GetXrefPropertyValue(string propertyName)
...
                 Monikers = Monikers,
                 Href = Href,
             };
+
+            // DHS appends branch infomation from cookie cache to URL, which is wrong for UID resolved URL
+            // output xref map with URL appending "?branch=master" for master branch
+            if (forOutput)
+            {
+                var (_, query, fragment) = UrlUtility.SplitUrl(Href);
+                var path = DeclairingFile.Docset.Repository.Branch == "master"
```
Comment (yufeih) [inline src/docfx/build/xref/InternalXrefSpec.cs]: `Docset.Repository` may be `null` when the docset is not managed by a git repository.

File: src/docfx/build/xref/InternalXrefSpec.cs [hunk 6]
```diff
@@ -29,19 +29,36 @@ public string GetXrefPropertyValue(string propertyName)
 
         public string GetName() => GetXrefPropertyValue("name");
 
-        public ExternalXrefSpec ToExternalXrefSpec(Context context, Document file)
+        public ExternalXrefSpec ToExternalXrefSpec(Context context, Document file, bool forOutput = true)
```
Comment (yufeih) [inline src/docfx/build/xref/InternalXrefSpec.cs]: should we add `branch=master` regardless? the only different is whether to output relative url or absolute url

File: src/docfx/build/xref/XrefMap.cs [hunk 7]
```diff
@@ -150,7 +139,16 @@ private IEnumerable<ExternalXrefSpec> ExpandInternalXrefSpecs()
                 if (TryGetValidXrefSpecs(uid, specsWithSameUid.Select(x => x.Value).ToList(), out var validInternalSpecs))
                 {
                     var internalSpec = GetLatestInternalXrefMap(validInternalSpecs);
-                    loadedInternalSpecs.Add((internalSpec as InternalXrefSpec).ToExternalXrefSpec(_context, internalSpec.DeclairingFile));
+                    var externalSpec = (internalSpec as InternalXrefSpec).ToExternalXrefSpec(_context, internalSpec.DeclairingFile);
+
+                    // DHS appends branch infomation from cookie cache to URL, which is wrong for UID resolved URL
+                    // output xref map with URL appending "?branch=master" for master branch
+                    var (path, query, fragment) = UrlUtility.SplitUrl(externalSpec.Href);
+                    path = internalSpec.DeclairingFile.Docset.Repository.Branch == "master"
```
[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: outdated, moved to `InternalXrefSpec`

File: src/docfx/build/xref/InternalXrefSpec.cs (hunk 6 above)
Comment (live1206) [reply src/docfx/build/xref/InternalXrefSpec.cs]: do you mean we should append `?branch=master` to internal UID resolved URL as well? For internal UID, the relative URL should exist with the current branch, no need to fall back to `master`, right?

**Pull request merged**